luskan/src/main/java/org/opencadc/luskan/TempStorageManager.java

```diff
@@ -250,4 +259,35 @@ public Content accept(String name, String contentType, InputStream inputStream)
     private static String getRandomString() {
         return new RandomStringGenerator(16).getID();
     }
+
+    private long getJobDestructionDelay() {
+        return this.job.getDestructionTime().getTime() - System.currentTimeMillis();
+    }
+
+    void scheduleDeletion(final File file, final long delay) {
```

Comment:
Does this actually work? the method creates a new executor on every call and schedules one job in there... then it calls shutdown() on the executor. So either:
* the scheduled job never runs, or 
* the shutdown() call blocks for a long time, or
* that executor API is really misleading because this looks wrong

Anyway, even if it worked you don't want an idle thread sitting there for ~2 days for every async query and inline upload.

Since there is a ScheduledExecutorService with a schedule(...) method, this code really needs a singleton: that means create one and store it for re-use. Since instances of TempStorageManager get created by every request, you need a static or lazy initialiser that creates the ScheduledExecutorService and stores it for later use. I would put the lazy init in the scheduleDeletion method. Then the body of the method simply gets the stored executor and schedules another deletion... the executor can run forever (until tomcat shuts down). 

Store it could be as a JNDI resource or as a static variable in TempStorageManager. The tricky bit is that you have no local way to come up with a name for the JNDI resource that is unique to this one application (on the grounds that this class could be used in two different apps for some open source reason) but I'm not sure that matters - if the code named the resource "TempStorageManager/deletion-executor" then all apps would share an executor and get it to delete files for them... and if it is safe, then just having 
```
private static ScheduledExecutorService SCHEDULER = Executors.newSingleThreadScheduledExecutor();
```
would work the same way and also be fine. The only catch is whether or not the thing will shutdown cleanly or not (ie is it a normal or daemon thread?) whent he JVM tries to shutdown. Can you see any reason that's a bad idea?

Reply:
Yes it does work. But yes it would also create X number of long lived threads.
Storing the scheduler service in the context would allow it to be retrieved and properly shutdown when the JVM terminates, but is the extra complexity worth it? A static shared among instances is simplest. Scheduled deletions not yet run could stop the jvm from exiting. From testing it does not stop the container from exiting. Using a ThreadFactory that returns daemon threads for the ScheduledExecutorService would allow the JVM to exit immediately. Deletion threads that happen to be running would be abruptly terminated, but for temp files I think that should be okay.